samples/third-party-lib/src/components/ThreeDimensionalGraph/ThreeDimensionalGraphModel.ts

```diff
@@ -0,0 +1,226 @@
+import { Feature } from "@vertigis/arcgis-extensions/data/Feature";
+import { FeatureStream } from "@vertigis/arcgis-extensions/data/FeatureStream";
+import { TableExtension } from "@vertigis/arcgis-extensions/data/TableExtension";
+import { QueryService } from "@vertigis/arcgis-extensions/tasks/query/QueryService";
+import {
+    ComponentModelBase,
+    serializable,
+    PropertyDefs,
+    ComponentModelProperties,
+} from "@vertigis/web/models";
+import { BrandingService } from "@vertigis/web/branding/BrandingService";
+import { command, Features } from "@vertigis/web/messaging";
+import { inject, FrameworkServiceType } from "@vertigis/web/services";
+import type {
+    GraphData as ForceGraphData,
+    LinkObject,
+    NodeObject as ForceNodeObject,
+} from "force-graph";
+import { toFeatureArray } from "./utils";
+
+interface ThreeDimensionalGraphModelProperties
+    extends ComponentModelProperties {
+    surveyTableExtension?: TableExtension;
+}
+
+export interface SurveyorNodeObject extends ForceNodeObject {
+    id: string;
+    type: "surveyor";
+}
+
+export interface SurveyNodeObject extends ForceNodeObject {
+    id: number;
+    type: "survey";
+    date: number;
+    hydrantNum: string;
+    result: "PASS" | "FAIL";
+}
+
+export type NodeObject = SurveyorNodeObject | SurveyNodeObject;
+
+export interface GraphData extends ForceGraphData {
+    nodes: NodeObject[];
+    links: LinkObject[];
+}
+
+@serializable
+export default class ThreeDimensionalGraphModel extends ComponentModelBase<
+    ThreeDimensionalGraphModelProperties
+> {
+    @inject(FrameworkServiceType.BRANDING)
+    brandingService: BrandingService | undefined;
+
+    // This will be populated from the app config item reference.
+    surveyTableExtension: TableExtension | undefined;
+
+    selectedSurveyId: number | undefined;
+    hydrants: Record<string, Feature> = {};
+    surveys: Record<number, Feature> = {};
+    graphData: GraphData = { links: [], nodes: [] };
+
+    getNodeLabel = (node: NodeObject): string => {
```

Comment:
Does the chart library sanitize HTML? If not, then you'll need to run it through `sanitizeHtml()` (import from "@vertigis/web/ui"). Either way, there needs to be a comment here explaining that using string concatenation to produce HTML is extremely dangerous in general. In this case it could be possible for someone to add features to the layer that include XSS in attribute data.

Reply:
Good catch. I've updated to sanitize/strip the html.

Reply:
Thanks, comment is perfect.